portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -17,50 +17,57 @@
 import com.google.appengine.api.datastore.DatastoreService;
 import com.google.appengine.api.datastore.DatastoreServiceFactory;
 import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.KeyFactory;
 import com.google.appengine.api.datastore.PreparedQuery;
 import com.google.appengine.api.datastore.Query;
 import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
 import com.google.gson.Gson;
 import java.io.IOException;
-import java.util.ArrayList;
 import java.util.List;
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.HashMap;
 import java.util.logging.Logger;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that allows users to comment on the portfolio. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
   private final static Logger LOGGER = Logger.getLogger(DataServlet.class.getName());
+  private DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+  private UserService userService = UserServiceFactory.getUserService();
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    // Get all the comments from the database
+    // Get all the comments from the database.
     Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
-
-    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
     PreparedQuery results = datastore.prepare(query);
 
-    List<String> comments = new ArrayList<String>();
+    Map<String, List<String>> comments = new HashMap<>();
     for (Entity entity : results.asIterable()) {
-      String comment = (String) entity.getProperty("comment");
-      comments.add(comment);
+      String id = KeyFactory.keyToString(entity.getKey());
+      List<String> commentEmail = new ArrayList<>();
+      commentEmail.add((String) entity.getProperty("comment"));
+      commentEmail.add((String) entity.getProperty("userEmail"));
+      comments.put(id, commentEmail);
```

Comment:
Why are we placing different types of attributes in a string list?
This forces the client to depend on the ordering of elements in a list, and makes the code fragile to change.
What alternative representation would be better?

Reply:
I am now using a map to store the information.